nanogpt/nanogpt_runner.cpp

```diff
@@ -0,0 +1,63 @@
+#include <cstdint>
+#include <functional>
+#include <memory>
+#include <unordered_map>
+
+#include "basic_tokenizer.h"
+#include "managed_tensor.h"
+
+#include <executorch/extension/module/module.h>
+#include <executorch/extension/evalue_util/print_evalue.h>
+#include <executorch/runtime/core/exec_aten/exec_aten.h>
+#include <executorch/runtime/core/exec_aten/util/dim_order_util.h>
+#include <executorch/runtime/core/exec_aten/util/tensor_util.h>
+
+using namespace std;
+using namespace torch::executor;
+
+using SizesType = exec_aten::SizesType;
+using DimOrderType = exec_aten::DimOrderType;
+using StridesType = exec_aten::StridesType;
+
+const int maxSeqLen = 1024;
+
+Result<vector<int64_t>> generate(unique_ptr<Module>& llm_model, vector<int64_t>& tokens) {
+    ManagedTensor tensor_tokens(tokens.data(), {1, 3}, ScalarType::Long);
+
+    vector<EValue> inputs;
+    inputs.push_back(tensor_tokens.get_aliasing_tensor());
+
+    Result<vector<EValue>> output_res = llm_model->forward(inputs);
+
+    Tensor output_tensor = output_res.get()[0].toTensor();
+    vector<int64_t> output_data(output_tensor.data_ptr<int64_t>(), output_tensor.data_ptr<int64_t>() + output_tensor.numel());
+
+    return output_data;
+}
+
+
+int main() {
+    // 1. get input
+    string prompt = "Hello world!";
+    cout << "prompt: " << prompt << endl;
+
+    // 2. load tokenizer, model and sampler.
+    BasicTokenizer tokenizer("local_vocab.json");
+
+    unique_ptr<Module> llm_model = make_unique<Module>(
+          /*model_path=*/"nanogpt.pte",
+          Module::MlockConfig::UseMlockIgnoreErrors);
+
+    // 3. tokenize the input
+    vector<int64_t> tokens = tokenizer.encode(prompt);
+
+    // 4. generate outputs
+    Result<vector<int64_t>> outputs = generate(llm_model, tokens);
+
+    // 5. decode the outputs
```

Comment:
Please expand on these one-line comments. "decode the outputs" is basically repeating what the code already says. But in this case we want to assume that people don't have a ton of experience with ML or LLMs, so this comment could discuss the fact that the model emits ids that map to English words in the tokenizer's map, and this step converts those IDs to words.